app/controllers/Giraffe.scala

```diff
@@ -124,8 +126,14 @@ class Giraffe(paymentServices: PaymentServices) extends Controller with Redirect
     val maxAmountInLocalCurrency = MaxAmount.forCurrency(countryGroup.currency)
     val creditCardExpiryYears = CreditCardExpiryYears(LocalDate.now.getYear, 10)
 
-    Ok(views.html.giraffe.contribute(pageInfo, maxAmountInLocalCurrency, countryGroup, chosenVariants, cmp, intCmp, creditCardExpiryYears, errorMessage))
-      .withCookies(chosenVariants.variants.map(Test.createCookie):_*)
+    val response = Ok(views.html.giraffe.contribute(pageInfo, maxAmountInLocalCurrency, countryGroup, variant, cmp, intCmp, creditCardExpiryYears, errorMessage))
+
+    val responseWithTests = (for {
+      testId <- request.cookies.get(Test.TestIdCookieName)
+      variantSlug <- request.cookies.get(Test.cookieName(variant))
+    } yield response) getOrElse response.withCookies(Test.testIdCookie(mvtId), Test.variantCookie(variant))
+
+    responseWithTests.discardingCookies(Test.allTests.filterNot(t => t.slug == variant.testSlug) map(t => DiscardingCookie(Test.cookieName(t))): _*)
```

Comment:
I've had a play about with this, but I don't know if we should discard all of the cookies first. The result of that is sending the client a `set-cookie` header on every response, even if that cookie is for the same test as the one they already had set. In that case, there's no need to do the `set-cookie`, you can just let the client keep the cookie as-is.  

It seems inefficient to constantly send `set-cookie` headers instead of only sending them when they're needed, which is how it works currently. What do you think?